### ISSUE TITLE
Tests failed on Kali Linux 2019.3 and segfault in test

### ISSUE THREAD

**Issue by mjptree** [shadow/shadow#699]
**Describe the issue**
Test project `/shadow/build ` fails with 18 tests out of 71 on Kali Linux, on Shadow master v1.13.1 on latest commit 43daab6. The tutorial test run as described in the Shadow documentation then fails with Segfault. Following tests failed:
```The following tests FAILED:
	  2 - shadow-leakcheck-compare (Failed)
	 22 - bind-shadow (SEGFAULT)
	 33 - epoll-writeable-shadow (SEGFAULT)
	 45 - shutdown-shadow (SEGFAULT)
	 51 - sockbuf-shadow (SEGFAULT)
	 53 - tcp-blocking-loopback-shadow (SEGFAULT)
	 54 - tcp-blocking-lossless-shadow (SEGFAULT)
	 55 - tcp-blocking-lossy-shadow (SEGFAULT)
	 57 - tcp-nonblocking-poll-loopback-shadow (SEGFAULT)
	 58 - tcp-nonblocking-poll-lossless-shadow (SEGFAULT)
	 59 - tcp-nonblocking-poll-lossy-shadow (SEGFAULT)
	 61 - tcp-nonblocking-epoll-loopback-shadow (SEGFAULT)
	 62 - tcp-nonblocking-epoll-lossless-shadow (SEGFAULT)
	 63 - tcp-nonblocking-epoll-lossy-shadow (SEGFAULT)
	 65 - tcp-nonblocking-select-loopback-shadow (SEGFAULT)
	 66 - tcp-nonblocking-select-lossless-shadow (SEGFAULT)
	 67 - tcp-nonblocking-select-lossy-shadow (SEGFAULT)
	 69 - tcp-iov-shadow (SEGFAULT)
```

**To Reproduce**
Followed tutorial instructions for building and installing shadow. Increased number of file descriptors allowed, as described in documentation. Then followed tutorial with example in `shadow/resource/examples`.

**Operating System (please complete the following information):**
 - OS and version: Kali GNU/Linus 2019.3
 - Kernel version: Linux kali 5.2.0-kali2-amd64 `#`1 SMP Debian 5.2.9-2kali1 (2019-08-22) x86_64 GNU/Linux

**Shadow (please complete the following information):**
 - Version: Shadow v1.13.1-3-g43daab65 2019-11-01 (built 2019-11-04) running GLib v2.62.2 and IGraph v0.7.1

 - Which plug-ins you are using:  
TGen

**Comment by robgjansen** [comment]
Thanks for reporting! This works for me on Cent OS, so it must be something specific to kali linux?

Since so many tests are failing, it is likely that something in the build process failed. Did you notice any errors from CMake during the `setup build` step?

After running the test, could you please check the output in `build/Testing/Temporary/LastTest.log` to see if any errors are apparent in that file for the failed test cases?

You can also check the Shadow log file for each test case by looking in the directory of one of the failed tests here `build/src/test/`.

**Comment by mjptree** [comment]
Forgot to mention, I should say: Running on VirtualBox 6.0.14. But yes, indeed a couple of compiler warnings were emitted - rebuilt with `./setup build -c -v -o --loader-valgrind --test`.
In file `shd-tcp-retransmit-tally.cc`:
```
~/shadow/src/main/host/descriptor/shd-tcp-retransmit-tally.cc: In member function ‘RetransmitTally& RetransmitTally::operator=(RetransmitTally&&)’:
~/shadow/src/main/host/descriptor/shd-tcp-retransmit-tally.cc:297:1: warning: no return statement in function returning non-void [-Wreturn-type]
  296 |    lost_ = std::move(rhs.lost_);
  +++ |+   return *this;
  297 | }
      | ^
```
In file `shd-process.c`:
```
~/shadow/src/main/host/shd-process.c: In function ‘_process_emu_selectHelper’:
~/shadow/src/main/host/shd-process.c:1648:12: warning: memmove
 1648 |             g_memmove(tmpReadFDs, readfds, sizeof(fd_set));
      |            ^~~~~~~~~
~/shadow/src/main/host/shd-process.c:1654:12: warning: memmove
 1654 |             g_memmove(tmpWriteFDs, writefds, sizeof(fd_set));
      |            ^~~~~~~~~
~/shadow/src/main/host/shd-process.c:1660:12: warning: memmove
 1660 |             g_memmove(tmpExceptFDs, exceptfds, sizeof(fd_set));
      |            ^~~~~~~~~
~/shadow/src/main/host/shd-process.c:1693:12: warning: memmove
 1693 |                     g_memmove(tmpReadFDs, readfds, sizeof(fd_set));
      |            ^        
~/shadow/src/main/host/shd-process.c:1697:12: warning: memmove
 1697 |                     g_memmove(tmpWriteFDs, writefds, sizeof(fd_set));
      |            ^        
~/shadow/src/main/host/shd-process.c:1701:12: warning: memmove
 1701 |                     g_memmove(tmpExceptFDs, exceptfds, sizeof(fd_set));
      |            ^        
~/shadow/src/main/host/shd-process.c:1709:12: warning: memmove
 1709 |             g_memmove(readfds, tmpReadFDs, sizeof(fd_set));
      |            ^~~~~~~~~
~/shadow/src/main/host/shd-process.c:1713:12: warning: memmove
 1713 |             g_memmove(writefds, tmpWriteFDs, sizeof(fd_set));
      |            ^~~~~~~~~
~/shadow/src/main/host/shd-process.c:1717:12: warning: memmove
 1717 |             g_memmove(exceptfds, tmpExceptFDs, sizeof(fd_set));
      |            ^~~~~~~~~
~/shadow/src/main/host/shd-process.c: In function ‘process_emu_readv’:
~/shadow/src/main/host/shd-process.c:2781:12: warning: memmove
 2781 |                         g_memmove(iov[i].iov_base, tempBuffer+bytesCopied, bytesToCopy);
      |            ^        
~/shadow/src/main/host/shd-process.c: In function ‘process_emu_writev’:
~/shadow/src/main/host/shd-process.c:2838:12: warning: memmove
 2838 |                     g_memmove(tempBuffer+bytesCopied, iov[i].iov_base, iov[i].iov_len);
      |            ^        
```
In file `shd-payload.c`:
```
~/shadow/src/main/routing/shd-payload.c: In function ‘payload_getData’:
~/shadow/src/main/routing/shd-payload.c:99:12: warning: memmove
   99 |         g_memmove(destBuffer, payload->data + offset, copyLength);
      |    
```
In files `shd-interposer.c`/`shd-preload-defs.h`:
```
In file included from ~/shadow/src/preload/shd-interposer.c:222:
/usr/include/pthread.h: In function ‘__pthread_unwind_next’:
~/shadow/src/preload/shd-preload-defs.h:281:1: warning: ‘noreturn’ function does return
  281 | PRELOADDEF(      , void, __pthread_unwind_next, (__pthread_unwind_buf_t *a), a);
      | ^
/usr/include/stdlib.h: In function ‘exit’:
~/shadow/src/preload/shd-interposer.c:417:1: warning: ‘noreturn’ function does return
  417 | }
      | ^
/usr/include/pthread.h: In function ‘pthread_exit’:
~/shadow/src/preload/shd-interposer.c:427:1: warning: ‘noreturn’ function does return
  427 | }
      | ^
/usr/include/stdlib.h: In function ‘abort’:
~/shadow/src/preload/shd-interposer.c:437:1: warning: ‘noreturn’ function does return
  437 | }
      | ^
```

A brief look at the test-log and I already see I think a null-dereference. I would clean up the log-file and upload it here, if that helps.

**Comment by mjptree** [comment]
@robgjansen I tried to rebuild it with debug symbols and had the same problems as initially, with 18 out of 71 tests failing. I then started shadow with gdb and tried to find a spot where the process jumps out of the rails, but it always terminated gracefully. Rerunning the tests, only test-sockbuf failed. But running test-sockbuf individually, it passes. So, I am unable to consistently reproduce the errors.
I have discussed with a friend and he means, it smells like a race condition somewhere. Will try to continue chasing the problem down...
